.github/workflows/release.yml

```diff
@@ -0,0 +1,115 @@
+    runs-on: ubuntu-latest
+    steps:
+      - name: Setup environment
+        run: |
+          set -eux
+          shopt -s extglob
+          EVENT_VER="${{ github.event.inputs.version }}"
+          if [[ $EVENT_VER ]]; then
+            RELEASE_VER=${EVENT_VER#v}
+          else
+            RELEASE_VER=${GITHUB_REF#refs/tags/v}
+          fi
+          if [[ $RELEASE_VER != +([0-9]).+([0-9]).+([0-9])* ]]; then
+            printf "Invalid version: %s\n" $RELEASE_VER
+            exit 1
+          fi
+          printf "RELEASE_VER=%s\n" ${RELEASE_VER} >> $GITHUB_ENV
+          printf "RELEASE_REF=refs/tags/v%s\n" ${RELEASE_VER} >> $GITHUB_ENV
+      - name: Checkout RNP release version
+        uses: actions/checkout@v2
+        with:
+          ref: ${{ env.RELEASE_REF }}
+          path: rnp-release
+      - name: Checkout latest RNP for signing
+        uses: actions/checkout@v2
+        with:
+          path: rnp-latest
+      - name: Checkout release-sign
+        uses: actions/checkout@v2
+        with:
+          repository: rnpgp/release-sign
+          path: release-sign
+      - name: Checkout Botan
+        uses: actions/checkout@v2
+        with:
+          repository: randombit/botan
+          ref: 2.17.3
+          path: botan
+      - name: Install RNP prerequisites
+        run: |
+          set -eux
+          sudo apt-get update
+          sudo apt-get -y install g++-8 cmake libbz2-dev zlib1g-dev libjson-c-dev build-essential gettext libncurses-dev
+          pushd "${GITHUB_WORKSPACE}/botan"
+          BOTAN_MODULES=$(<../rnp-release/ci/botan-modules tr '\n' ',')
+          ./configure.py --cxxflags="-fno-omit-frame-pointer" --without-documentation --without-openssl --build-targets=shared --minimized-build --enable-modules="$BOTAN_MODULES"
+          make -j2 -s
+          sudo make -j2 install
+          popd
+          sudo ldconfig /usr/local/lib
+      - name: Install RNP
+        run: |
+          set -eux
+          mkdir "${GITHUB_WORKSPACE}/rnp-build"
+          pushd "${GITHUB_WORKSPACE}/rnp-build"
+          cmake -DDOWNLOAD_RUBYRNP=Off -DBUILD_TESTING=Off ../rnp-release/
+          make -j2 -s
+          sudo make -j2 install
+          popd
+          rnp --version
+      - name: Sign sourceballs
+        env:
+          RELEASE_PUBLIC_KEY: ${{ secrets.RNPGP_RELEASE_SIGNING_PUBLIC_KEY }}
+          RELEASE_SECRET_KEY: ${{ secrets.RNPGP_RELEASE_SIGNING_SECRET_KEY }}
+        run: |
+          set -eux
+          pushd "${GITHUB_WORKSPACE}/release-sign"
+          mkdir ~/.rnp
+          set +x
+          if [[ -z $RELEASE_PUBLIC_KEY ]]; then
+            printf "RELEASE_PUBLIC_KEY is not set!\n"
+            exit 1
+          fi
+          if [[ -z $RELEASE_SECRET_KEY ]]; then
+            printf "RELEASE_SECRET_KEY is not set!\n"
+            exit 1
+          fi
+          printf "%s" "${RELEASE_PUBLIC_KEY}" > release-pub.asc
+          printf "%s" "${RELEASE_SECRET_KEY}" > release-sec.asc
```

Comment:
Thanks, this way would be definitely better! And now I know some more bash :) As I read this would still create some temporary file, which would be cleared but not wiped, I guess. So real solution with this behaviour would be to add some way for RNP to use data (including key input) from the environment variables.

Reply:
Alternatively, if we don't want the hassle of having to `shred` at a later stage, we can do the following:

```
./rel-sign.sh -r "${{ github.repository }}" \
  -v "${RELEASE_VER}" \
  --src "${GITHUB_WORKSPACE}"/rnp-release \
  --pparams \
  --keyfile <(cat "${RELEASE_SECRET_KEY}")
```

... and the process list will just look like this:

```
./rel-sign.sh -r ... -v ... --src ..../rnp-release --pparams --keyfile /dev/fd/63
```

Reply:
@ribose-jeffreylau I tried it in the way like this (via `/dev/stdin`), and here comes the problem - the signing occurs twice :-) I'm thinking on adding some special way to work with environment variables for rnp/rnpkeys - i.e. `--keyfile ":RELEASE_SECRET_KEY"` or `"env: RELEASE_SECRET_KEY"`. Are you aware of how this may be done in other CLIs, like `-` is common for `stdin`?

Reply:
@ni4 It is strange that it would sign twice, which sounds like `./rel-sign.sh` is being called twice?  Just clarifying that I meant replacing these:

```
<<<"${RELEASE_SECRET_KEY}" cat > release-sec.asc
```

```
./rel-sign.sh -r "${{ github.repository }}" \
  -v "${RELEASE_VER}" \
  --src "${GITHUB_WORKSPACE}"/rnp-release \
  --pparams \
  --keyfile release-sec.asc
```

with just this:

```
./rel-sign.sh -r "${{ github.repository }}" \
  -v "${RELEASE_VER}" \
  --src "${GITHUB_WORKSPACE}"/rnp-release \
  --pparams \
  --keyfile <(cat "${RELEASE_SECRET_KEY}")
```

Bash should then take care of creating a file descriptor pointing to buffer with content of `$RELEASE_SECRET_KEY`, and replacing the `<(...)` above with the file descriptor (`"/dev/fd/..."`).
I expect `rnpkeys` would receive this `/dev/fd/...` file handle and then read the secret key from it just like a normal file.

Answering your question, I believe `-` is a common argument for a "file name" meant to represent `stdin`.

And for the "special way to work with environment variables", I'm not sure if it is necessary -- will it add much complexity?  Are there alternatives that are much simpler / more established?  I'm wondering about these since I have no experience in reading from environment in a security software setting 😛  But if it's deemed fit to implement, I would make the option string explicit about reading from the environment, something like `--key-env-name "RELEASE_SECRET_KEY"` instead of overloading `--keyfile`.

Reply:
@ribose-jeffreylau 
>  It is strange that it would sign twice, which sounds like ./rel-sign.sh is being called twice?

I mean that, if we feed key to the script via stdin, then it is consumed by first `rnpkeys`, and for the second it is empty (since script calls signing twice - for `.tar.gz` and for `.zip`).

> Bash should then take care of creating a file descriptor pointing to buffer with content of $RELEASE_SECRET_KEY, and replacing the <(...) above with the file descriptor ("/dev/fd/...").
I expect rnpkeys would receive this /dev/fd/... file handle and then read the secret key from it just like a normal file.

Thanks for this! Now I know even more :) However, this doesn't work as well - looks like that tmp file is deleted right after it was read by the first `rnp --sign` call. I.e. first call succeeds, but second errors.

> And for the "special way to work with environment variables", I'm not sure if it is necessary -- will it add much complexity? Are there alternatives that are much simpler / more established?

As from the CLI side it would not be hard to implement, as reading is abstracted out, probably easier then fighting with bash :) Just add some special prefix (like `env:` or just `:`)  so CLI would understand that instead of reading from file with name it should read from memory, pointed by env variable. Or from the stdin, if `-` is specified.

Reply:
For your last point, the reason I asked is that I don't think it's straightforward to deal with files named `env:...`...

For the calling twice point, I think I get it now 😆 

I think here's what's happening:

```console
$ testbash() { local file="$1"; echo 1; wc -c "$file"; echo 2; wc -c "$file"; echo done.; }
$ testbash <(echo bash it)
1
       8 /dev/fd/63
2
       0 /dev/fd/63
done.
```

`testbash` is doing what `rel-sign.sh` is doing, and `wc -c`s are each call to sign.

My tendency is to do all the gluing work outside of `rnp` cli itself (i.e. in `rel-sign`), but doing it in the cli itself will have a higher chance of having tests written against it, which is a good thing.

Reply:
As an example of doing the glue work outside of `rnp`, see https://github.com/rnpgp/release-sign/pull/4

Reply:
@ribose-jeffreylau Understand now, it's really more logical to update scripts which use CLI then update CLI since it is needed in scripts. However, why not to give some flexibility to user as well, so they would not need to write complex scripts to automate some things?

Reply:
@ni4 Yes, that's fine - though my vote would be to create a new `--flag` that differentiates itself from `--keyfile`.

Reply:
@ribose-jeffreylau My idea was to make this universal for all path-related options and values, i.e. be able to load signatures from stdin/env, save signature there and so on.

Reply:
@ni4 Ah, OK!  Maybe belongs to another issue.